File: core/src/main/scala/diplomacy/variant/standard/rule/MovementResolver.scala
package diplomacy.variant.standard.rule

import scala.collection.{ mutable => cmutable }

import diplomacy.rule.{ ResolvedResult, OrderResult, InvalidOrderMessage }

class MovementResolver extends Rule.TypeHelper {
  def apply(
    board: Board, orders: Set[MovementOrder[Power]]
  ): Either[InvalidOrderMessage, ResolvedResult] = {
    val ordersWithResult = orders map { order => new mutable.MovementOrderWithResult[Power](order) }
    val movesViaConvoy = cmutable.Set[Order.Move[Power]]()
    val dislodgedFrom = cmutable.Map[DiplomacyUnit, Province]()

    def canBounce(
      order1: mutable.MovementOrderWithResult[Power],
      order2: mutable.MovementOrderWithResult[Power]
    ): Boolean = {
      (order1.order, order2.order) match {
        case (m1@ Order.Move(_, _, _), m2@ Order.Move(_, _, _)) =>
          if (m1.destination.province == m2.destination.province) {
            true
          } else if (
            (m1.destination.province == m2.unit.location.province) &&
            (m2.destination.province == m1.unit.location.province)
          ) {
            if ((movesViaConvoy contains m1) || (movesViaConvoy contains m2)) {
              order1.result == Some(Result.Bounced) || order2.result == Some(Result.Bounced)
            } else {
              true
            }
          } else if (m1.destination.province == m2.unit.location.province) {
            order2.result == Some(Result.Bounced)
          } else if (m2.destination.province == m1.unit.location.province) {
            order1.result == Some(Result.Bounced)
          } else {
            false
          }
        case (m1@ Order.Move(_, _, _), _) =>
          m1.destination.province == order2.order.unit.location.province
        case (_, m2@ Order.Move(_, _, _)) =>
          m2.destination.province == order1.order.unit.location.province
        case (_, _) =>
          false
      }
    }

    // 1. Divide orders into groups
    final class TmpMovementOrderGroup {
      type MovementOrderWithResult =
        diplomacy.variant.standard.rule.mutable.MovementOrderWithResult[Power]

      var target: Option[MovementOrderWithResult] = None
      val relatedOrders = cmutable.Set[MovementOrderWithResult]()
    }

    val province2TmpOrderGroups =
      cmutable.Map[Province, cmutable.Map[Location, TmpMovementOrderGroup]]()
    ordersWithResult foreach { order =>
      order.order match {
        case Order.Hold(unit) =>
          val groups =
            province2TmpOrderGroups.getOrElseUpdate(unit.location.province, cmutable.Map())
          val group =
            groups.getOrElseUpdate(unit.location, new TmpMovementOrderGroup())
          group.target = Option(order)
        case Order.Move(unit, destination, _) =>
          val groups1 =
            province2TmpOrderGroups.getOrElseUpdate(destination.province, cmutable.Map())
          val group1 =
            groups1.getOrElseUpdate(unit.location, new TmpMovementOrderGroup())
          group1.target = Option(order)

          val groups2 =
            province2TmpOrderGroups.getOrElseUpdate(unit.location.province, cmutable.Map())
          val group2 =
            groups2.getOrElseUpdate(unit.location, new TmpMovementOrderGroup())
          group2.target = Option(order)
        case s@ Order.Support(unit, target) =>
          val groups1 = province2TmpOrderGroups.getOrElseUpdate(
            s.destinationLocation.province, cmutable.Map()
          )
          val group1 = groups1.getOrElseUpdate(
            s.targetOrder.unit.location, new TmpMovementOrderGroup()
          )
          group1.relatedOrders += order

          val groups2 =
            province2TmpOrderGroups.getOrElseUpdate(unit.location.province, cmutable.Map())
          val group2 =
            groups2.getOrElseUpdate(unit.location, new TmpMovementOrderGroup())
          group2.target = Option(order)
        case Order.Convoy(unit, target) =>
          val groups1 = province2TmpOrderGroups.getOrElseUpdate(
            target.destination.province, cmutable.Map()
          )
          val group1 = groups1.getOrElseUpdate(
            target.unit.location, new TmpMovementOrderGroup()
          )
          group1.relatedOrders += order

          val groups2 =
            province2TmpOrderGroups.getOrElseUpdate(unit.location.province, cmutable.Map())
          val group2 =
            groups2.getOrElseUpdate(unit.location, new TmpMovementOrderGroup())
          group2.target = Option(order)
        case _ =>
      }
    }

    // 2. Exclude support or convoy orders that have no corresponding orders
    for {
      (province, groups) <- province2TmpOrderGroups
      (location, group) <- groups if group.target.isEmpty
    } {
      group.relatedOrders.foreach { order => order.result = Result.NoCorrespondingOrder }
      groups -= location
    }
    val province2OrderGroups = (province2TmpOrderGroups map {
      case (province, tmpGroups) =>
        val groups = tmpGroups flatMap {
          case (location, tmpGroup) =>
            tmpGroup.target map { targetOrder =>
              location -> new mutable.MovementOrderGroup(targetOrder, tmpGroup.relatedOrders.toSet)
            }
        }
        province -> groups
    }).toMap

    // 3. Generate the dependency graph
    var graph = new OrderDependency[Power](ordersWithResult map { _.toImmutable }).graph

    // 4. Resolve orders following dependency
    while (graph.vertices.size != 0) {
      val target =
        graph.vertices.find {
          node => graph.edges.forall { case (_, p1) => node != p1 }
        }

      // TODO
      target foreach { provinces =>
        graph = graph.remove(provinces)

        // Get all related order groups
        val relatedGroups = cmutable.Map() ++ (provinces map { province =>
          val groups =
            province2OrderGroups.getOrElse(province, cmutable.Map()) map { _._2 }

          province -> (cmutable.Set() ++ groups)
        })

        // Resolve each province
        while (relatedGroups.size != 0) {
          // 1. Resolve cutting support
          for {
            (_, groups) <- relatedGroups
            group <- groups
          } {
            group.target.order match {
              case s@ Order.Support(unit, target) =>
                val destination = s.destinationLocation
                val isCut = groups exists { group =>
                  group.target.order match {
                    case m@ Order.Move(unit, dest, _) =>
                      (unit.location.province != destination.province) &&
                      (group.route(board.map).isDefined) &&
                      (s.unit.power != m.unit.power)
                    case _ => false
                  }
                }

                if (isCut) {
                  group.target.result = Result.Cut
                }
              case _ =>
            }
          }

          // 2. Sort provinces by related units
          val sortedGroups =
            relatedGroups.toSeq sortWith {
              case ((_, g1), (_, g2)) =>
                val hasG1Convoy = g1 exists { _.target.order match {
                  case Order.Convoy(_, _) => true
                  case _ => false
                }}
                val hasG2Convoy = g2 exists { _.target.order match {
                  case Order.Convoy(_, _) => true
                  case _ => false
                }}

                if (hasG1Convoy) {
                  true
                } else if (hasG2Convoy) {
                  false
                } else {
                  val pow1 = (g1 map { group => group.power }).sum
                  val pow2 = (g2 map { group => group.power }).sum
                  pow1 > pow2
                }
              }

            // 3. Check whether move orders can be conducted or not
            val failedMoves = cmutable.Set[mutable.MovementOrderGroup[Power]]()
            for {
              (province, groups) <- sortedGroups
              group <- groups
            } {
              group.target.order match {
                case m@ Order.Move(_, destination, _) if province == destination.province =>
                  group.route(board.map) match {
                    case Some(route) =>
                      if (route.useConvoy) {
                        movesViaConvoy += m
                      }
                    case None =>
                      failedMoves += group
                  }
                case _ =>
              }
            }

          // 4. Resolve the province, and delete it from the buffer
          sortedGroups.headOption foreach {
            case (province, groups) =>
              relatedGroups -= province

              // 5. Resolve and exclude failed move orders
              groups foreach { group =>
                if (failedMoves contains group) {
                  group.target.result = Result.Failed
                  group.relatedOrders foreach { o => o.result = Result.Failed }
                  groups -= group
                }
              }

              val defenceOpt = groups find { group =>
                group.target.order.unit.location.province == province
              }
              val offence = groups -- defenceOpt

              // 5. Resolve and exclude dislodged moves if #provinces <= 2
              if (provinces.size <= 2) {
                defenceOpt foreach { defence =>
                  offence foreach { group =>
                    group.target.order match {
                      case m @Order.Move(unit, location, _) =>
                        val isDislodged = group.target.result map {
                          case Result.Dislodged(_) => true
                          case _ => false
                        } getOrElse false
                        if (isDislodged && canBounce(group.target, defence.target)) {
                          groups -= group
                          offence -= group
                          group.relatedOrders foreach { o => o.result = Result.Failed }
                        }
                      case _ =>
                    }
                  }
                }
              }

              if (groups.nonEmpty) {
                // 6. Find orders that have the highest power
                val maxPower = (groups maxBy { _.power }).power
                val maxOrders = groups withFilter { _.power >= maxPower } map { _.target }

                // 7. Resolve the defence order
                defenceOpt foreach { defence =>
                  if (maxOrders contains defence.target) {
                    defence.target.order match {
                      case Order.Move(_, _, _) =>
                      case Order.Hold(_) =>
                        defence.target.result = Result.Success
                        defence.relatedOrders foreach { _.result = Result.Success }
                      case Order.Convoy(_, _) =>
                        defence.target.result match {
                          case Some(Result.NoCorrespondingOrder) =>
                          case _ =>
                            defence.target.result = Result.Failed // This convoy order is available.
                        }
                        defence.relatedOrders foreach { _.result = Result.Success }
                      case _ =>
                        defence.relatedOrders foreach { _.result = Result.Success }
                    }
                  } else {
                    var isDislodged = false
                    val offenceGroup =
                      offence find { group => maxOrders contains group.target }
                    if (maxOrders.size == 1) {
                      // Check self dislodgement
                      offenceGroup foreach { group =>
                        var offensivePower = 0
                        if (group.target.order.unit.power != defence.target.order.unit.power) {
                          offensivePower += 1
                        }

                        val validSupports = group.validSupports filter {
                          _.unit.power != defence.target.order.unit.power
                        }
                        offensivePower += validSupports.size

                        isDislodged = defence.power < offensivePower
                      }
                    }

                    if (isDislodged) {
                      offenceGroup foreach { group =>
                        defence.target.result =
                          Result.Dislodged(group.target.order.unit.location.province)
                        dislodgedFrom(defence.target.order.unit) =
                          group.target.order.unit.location.province
                        defence.relatedOrders foreach { o => o.result = Result.Failed }
                      }
                    } else {
                      defence.target.order match {
                        case h@ Order.Hold(_) =>
                          defence.target.result = Result.Success
                          defence.relatedOrders foreach { o => o.result = Result.Success }
                        case _ =>
                      }
                    }
                  }
                }

                // 8. Resolve the offence orders
                offence foreach { group =>
                  var isBounced = true
                  if (maxOrders contains group.target) {
                    if (maxOrders.size == 1) {
                      defenceOpt match {
                        case Some(defence) =>
                          isBounced =
                            defence.target.result map {
                              case Result.Dislodged(_) => false
                              case _ => true
                            } getOrElse true
                        case None => isBounced = false
                      }
                    } else if (maxOrders.size == 2){
                      val order2 = (maxOrders find { _ != group.target }).get
                      isBounced = canBounce(group.target, order2)
                    }
                  }

                  if (isBounced) {
                    group.target.result = Result.Bounced
                    group.relatedOrders foreach { o => o.result = Result.Failed }
                  } else {
                    group.target.result = Result.Success
                    group.relatedOrders foreach { o => o.result = Result.Success }
                  }
                }
              }
          }
        }
      }
    }

    // Generate a new board
    val unit2Result = (ordersWithResult map { order =>
      order.order.unit -> (order.order, order.result)
    }).toMap

    val newUnits: Set[DiplomacyUnit] =
      board.units map { unit =>
        unit2Result.get(unit) match {
          case Some((order, result)) =>
            order match {
              case m@ Order.Move(_, _, _) if result == Option(Result.Success) =>
                unit.copy(location = m.destination)
              case _ => unit
            }
          case None => unit
        }
      }

    val newUnitStatuses: Map[DiplomacyUnit, UnitStatus] =
      (ordersWithResult flatMap { order =>
        order.result match {
          case Some(Result.Dislodged(_)) =>
            dislodgedFrom.get(order.order.unit) map { province =>
              order.order.unit -> UnitStatus.Dislodged(province)
            }
          case _ => None
        }
      }).toMap

    val provincesContainingUnit = newUnits map { _.location.province }
    val newProvinceStatuses: Map[Province, ProvinceStatus] =
      province2OrderGroups flatMap {
        case (province, groups) =>
          val wasBounced =
            groups exists { case (_, g) => g.target.result == Option(Result.Bounced) }
          if (
            !(provincesContainingUnit contains province) &&
            wasBounced
          ) {
            Option(province -> ProvinceStatus.Standoff)
          } else {
            None
          }
      }

    val newState = board.state.copy(phase = Retreat)

    val newBoard = board.copy(
      state = newState,
      units = newUnits.toSet,
      unitStatuses = newUnitStatuses,
      provinceStatuses = newProvinceStatuses
    )

    val orderResults: Set[OrderResult] =
      ordersWithResult flatMap { order =>
        order.result map { result =>
          OrderResult.Executed[Power, MilitaryBranch, Order, Result](order.order, result)
        }
      }
    Right(ResolvedResult(newBoard, orderResults))
  }
}

File: src/variant/standard/rule/movement-resolver.ts
import { Resolver } from "./resolver"
import { Result } from "./result"
import { Phase } from "./phase"
import { State } from "./state"
import { OrderType, Order, Hold, Move, Support, Convoy } from "./order"
import { MilitaryBranch } from "./military-branch"
import { Dislodged } from "./dislodged"
import { ProvinceStatus } from "./province-status"
import { Board, Location, Unit, ResolvedResult } from "./types"
import { MovementOrderWithResult } from "./movement-order-with-result"
import { MovementOrderGroup } from "./movement-order-group"
import { OrderDependency } from "./order-dependency"
import { Province } from "./../../../board/module"
import { Executed } from "./../../../rule/module"
import { Success } from "./../../../util/module"

const { Retreat } = Phase

class TmpMovementOrderGroup<Power> {
  constructor (
    public target: MovementOrderWithResult<Power> | null,
    public relatedOrders: Set<MovementOrderWithResult<Power>>
  ) {}
}


export class MovementResolver<Power> implements Resolver<Power> {
  resolve (board: Board<Power>, orders: Set<Order<Power>>) {
    const ordersWithResult = [...orders].map(order => new MovementOrderWithResult(order))
    const movesViaConvoy = new Set()
    const dislodgedFrom = new Map()

    function canBounce (
      order1: MovementOrderWithResult<Power>, order2: MovementOrderWithResult<Power>
    ) {
      const o1 = order1.order
      const o2 = order2.order
      if (o1 instanceof Move && o2 instanceof Move) {
        if (o1.destination.province === o2.destination.province) {
          return true
        } else if (
          (o1.destination.province === o2.unit.location.province) &&
          (o2.destination.province === o1.unit.location.province)
        ) {
          if (movesViaConvoy.has(order1.order) || movesViaConvoy.has(order2.order)) {
            return order1.getResult() === Result.Bounced || order2.getResult() === Result.Bounced
          } else {
            return true
          }
        } else if (o1.destination.province === o2.unit.location.province) {
          return order2.getResult() === Result.Bounced
        } else if (o2.destination.province === o1.unit.location.province) {
          return order1.getResult() === Result.Bounced
        } else {
          return false
        }
      } else if (o1 instanceof Move) {
        return o1.destination.province === order2.order.unit.location.province
      } else if (o2 instanceof Move) {
        return o2.destination.province === order1.order.unit.location.province
      } else {
        return false
      }
    }

    // 1. Divide orders into groups
    const province2TmpOrderGroup = new Map()
    function getOrderGroup (province: Province<Power>, location: Location<Power>) {
      const groups = province2TmpOrderGroup.get(province) || new Map()
      const group = groups.get(location) || new TmpMovementOrderGroup(null, new Set())
      groups.set(location, group)
      province2TmpOrderGroup.set(province, groups)

      return group
    }
    ordersWithResult.forEach(order => {
      const group = getOrderGroup(order.order.unit.location.province, order.order.unit.location)
      group.target = order
      const o = order.order
      if (o instanceof Hold) {
      } else if (o instanceof Move) {
        const o2: Move<Power> = o
        const group2 = getOrderGroup(o2.destination.province, o2.unit.location)
        group2.target = order
      } else if (o instanceof Support) {
        const o2: Support<Power> = o
        const group2 =
          getOrderGroup(o2.destination.province, o2.target.unit.location)
        group2.relatedOrders.add(order)
      } else if (o instanceof Convoy) {
        const o2: Convoy<Power> = o
        const group2 =
          getOrderGroup(o2.target.destination.province, o2.target.unit.location)
        group2.relatedOrders.add(order)
      }
    })
    // 2. Exclude support or convoy orders that have no corresponding orders
    for (let elem of [...province2TmpOrderGroup]) {
      for (let elem2 of [...elem[1]]) {
        if (!elem2[1].target) {
          elem2[1].relatedOrders.forEach((order: MovementOrderWithResult<Power>) => {
            order.setResult(Result.NoCorrespondingOrder)
          })
          elem[1].delete(elem2[0])
        }
      }
    }
    const province2OrderGroups = new Map<Province<Power>, Set<MovementOrderGroup<Power>>>()
    province2TmpOrderGroup.forEach((groups, province) => {
      const newGroups = [...groups.values()].map(x => {
        return new MovementOrderGroup<Power>(x.target, x.relatedOrders)
      })
      province2OrderGroups.set(province, new Set([...newGroups]))
    })

    // 3. Generate the dependency graph
    let graph = new OrderDependency(ordersWithResult).graph

    // 4. Resolve orders following dependency
    while (graph.nodes.size > 0) {
      const target = [...graph.nodes].find(node => {
        return [...graph.edges].every(elem => elem.n1 !== node)
      })

      if (!target) {
        throw "Internal Error"
      }

      graph = graph.deleteNode(target)
      const provinces = target

      // Get all related order groups
      const relatedGroups = new Map([...provinces].map((province: Province<Power>) => {
        const groups = province2OrderGroups.get(province) || new Set()
        return <[Province<Power>, Set<MovementOrderGroup<Power>>]>([province, new Set([...groups])])
      }))

      // Resolve each province
      while (relatedGroups.size !== 0) {
        // 1. Resolve cutting support
        relatedGroups.forEach(groups => {
          groups.forEach(group => {
            const o = group.target.order
            if (o instanceof Support) {
              const destination = o.destination
              const isCut = [...groups].some(group => {
                const o2 = group.target.order
                if (o2 instanceof Move) {
                  return (
                    (group.target.order.unit.location.province !== destination.province) &&
                    (group.route(board.map)) &&
                    (o.unit.power !== group.target.order.unit.power)
                  ) || false
                }
                return false
              })
              if (isCut) {
                group.target.setResult(Result.Cut)
              }
            }
          })
        })

        // 2. Sort provinces by related units
        const sortedGroups = [...relatedGroups].sort((a, b) => {
          const hasG1Convoy = [...a[1]].some(group => {
            return group.target.order.tpe === OrderType.Convoy
          })
          const hasG2Convoy = [...b[1]].some(group => {
            return group.target.order.tpe === OrderType.Convoy
          })

          if (hasG1Convoy) {
            return -1
          } else if (hasG2Convoy) {
            return 1
          } else {
            const pow1 = Math.max(...[...a[1]].map(group => group.power()))
            const pow2 = Math.max(...[...b[1]].map(group => group.power()))
            return (pow1 > pow2) ? -1 : 1
          }
        })

        // 3. Check whether move orders can be conducted or not
        const failedMoves = new Set()
        sortedGroups.forEach(elem => {
          const [province, groups] = elem
          groups.forEach(group => {
            const o = group.target.order
            if (
              (o instanceof Move) &&
              (province === o.destination.province)
            ) {
              const route = group.route(board.map)
              if (route) {
                if (route.viaConvoy) {
                  movesViaConvoy.add(group.target.order)
                }
              } else {
                failedMoves.add(group)
              }
            }
          })
        })

        if (sortedGroups.length === 0) {
          continue
        }

        // 4. Resolve the province, and delete it from the buffer
        const [province, groups] = sortedGroups[0]
        relatedGroups.delete(province)

        // 5. Resolve and exclude failed move orders
        groups.forEach(group => {
          if (failedMoves.has(group)) {
            group.target.setResult(Result.Failed)
            group.relatedOrders.forEach(o => { o.setResult(Result.Failed) })
            groups.delete(group)
          }
        })

        const defenceOpt = [...groups].find(group => {
          return group.target.order.unit.location.province === province
        })
        const offence = new Set([...groups])
        if (defenceOpt) offence.delete(defenceOpt)

        // 6. Resolve and exclude dislodged moves if #provinces <= 2
        if ([...provinces].length <= 2) {
          if (defenceOpt) {
            offence.forEach(group => {
              if (group.target.order.tpe === OrderType.Move) {
                const isDislodged =
                  (!group.target.getResult()) ? false : group.target.getResult() === Result.Dislodged
                if (isDislodged && canBounce(group.target, defenceOpt.target)) {
                  groups.delete(group)
                  offence.delete(group)
                  group.relatedOrders.forEach(o => { o.setResult(Result.Failed) })
                }
              }
            })
          }
        }

        if (groups.size === 0) continue

        // 7. Find orders that have the highest power
        const maxPower = Math.max(...[...groups].map(group => group.power()))
        const maxOrders =
          new Set([...groups].filter(g => g.power() >= maxPower).map(g => g.target))

        // 8. Resolve the defence order
        if (defenceOpt) {
          if (maxOrders.has(defenceOpt.target)) {
            const o = defenceOpt.target.order
            if (o instanceof Move) {

            } else if (o instanceof Hold) {
              defenceOpt.target.setResult(Result.Success)
              defenceOpt.relatedOrders.forEach(o => { o.setResult(Result.Success) })
            } else if (o instanceof Convoy) {
              if (defenceOpt.target.getResult() !== Result.NoCorrespondingOrder) {
                defenceOpt.target.setResult(Result.Failed) // This convoy order is available.
              }
              defenceOpt.relatedOrders.forEach(o => { o.setResult(Result.Success) })
            } else {
                defenceOpt.relatedOrders.forEach(o => { o.setResult(Result.Success) })
            }
          } else {
            let isDislodged = false
            const offenceGroup = [...offence].find(group => maxOrders.has(group.target))
            if (maxOrders.size === 1) {
              if (offenceGroup) {
                let offensivePower = 0
                if (offenceGroup.target.order.unit.power !== defenceOpt.target.order.unit.power) {
                  offensivePower += 1
                }
                const validSupports = [...offenceGroup.validSupports()].filter(s => {
                  return s.unit.power !== defenceOpt.target.order.unit.power
                })
                offensivePower += validSupports.length

                isDislodged = defenceOpt.power() < offensivePower
              }

              if (isDislodged) {
                if (offenceGroup) {
                  defenceOpt.target.setResult(Result.Dislodged)
                  dislodgedFrom.set(
                    defenceOpt.target.order.unit, offenceGroup.target.order.unit.location.province
                  )
                  defenceOpt.relatedOrders.forEach(order => { order.setResult(Result.Failed) })
                }
              } else {
                if (defenceOpt.target.order.tpe === OrderType.Hold) {
                  defenceOpt.target.setResult(Result.Success)
                  defenceOpt.relatedOrders.forEach(o => { o.setResult(Result.Success) })
                }
              }
            }
          }
        }

        // 9. Resolve the offence orders
        offence.forEach(group => {
          let isBounced = true
          if (maxOrders.has(group.target)) {
            if (maxOrders.size === 1) {
              if (defenceOpt) {
                if (defenceOpt.target.order.tpe === OrderType.Move) {
                  isBounced = (defenceOpt.target.getResult() !== Result.Success) &&
                    (defenceOpt.target.getResult() !== Result.Dislodged)
                } else {
                  isBounced = defenceOpt.target.getResult() !== Result.Dislodged
                }
              } else {
                isBounced = false
              }
            } else if (maxOrders.size === 2) {
              const order2 =
                <MovementOrderWithResult<Power>>([...maxOrders].find(o => o !== group.target))
              isBounced = canBounce(group.target, order2)
            }
          }
          if (isBounced) {
            group.target.setResult(Result.Bounced)
            group.relatedOrders.forEach(o => { o.setResult(Result.Failed) })
          } else {
            group.target.setResult(Result.Success)
            group.relatedOrders.forEach(o => { o.setResult(Result.Success) })
          }
        })
      }
    }

    // Generate a new board
    const unit2Result = new Map(ordersWithResult.map(order => {
      return <[Unit<Power>, [Order<Power>, Result]]>[order.order.unit, [order.order, order.getResult()]]
    }))

    const newUnits = new Set()
    board.units.forEach(unit => {
      const r = unit2Result.get(unit)
      if (r) {
        const [order, result] = r
        if (order instanceof Move && result === Result.Success) {
          newUnits.add(new Unit(unit.militaryBranch, order.destination, unit.power))
        } else {
          newUnits.add(unit)
        }
      } else {
        newUnits.add(unit)
      }
    })
    const newUnitStatuses = new Map()
    ordersWithResult.forEach(order => {
      if (order.getResult() === Result.Dislodged) {
        const x = dislodgedFrom.get(order.order.unit)
        newUnitStatuses.set(order.order.unit, new Dislodged(x))
      }
    })
    const provincesContainingUnit = new Set([...newUnits].map(u => u.location.province))
    const occupationStatuses = [...board.provinceStatuses]
      .filter(elem => elem[1].occupied)
      .map(elem => {
        if (elem[1].standoff) {
          return <[Province<Power>, ProvinceStatus<Power>]>[elem[0], new ProvinceStatus(elem[1].occupied, false)]
        } else {
          return elem
        }
      })
    const newProvinceStatuses = new Map([...occupationStatuses])
    province2OrderGroups.forEach((groups, province) => {
      const wasBounced = [...groups].some(group => group.target.getResult() === Result.Bounced)
      const standoff = wasBounced && !(provincesContainingUnit.has(province))

      if (standoff) {
        const current = newProvinceStatuses.get(province) || new ProvinceStatus(null, true)
        const x = new ProvinceStatus(current.occupied, true)
        newProvinceStatuses.set(province, x)
      }
    })
    const newState = new State(board.state.turn, Retreat)
    const newBoard = new Board(board.map, newState, newUnits, newUnitStatuses, newProvinceStatuses)

    const orderResults = [...ordersWithResult].map(order => {
      const r = order.getResult()
      if (r !== null) {
        return new Executed<Power, MilitaryBranch, Result>(order.order, r)
      } else {
        throw `Internal error: ${order.order} was not resolved`
      }
    })

    return new Success(new ResolvedResult(newBoard, orderResults, false))
  }
}
